### PR TITLE
Migrate from `arrow` (`apache/arrow-rs`) to `arrow2` (unofficial, safer)

### PR DESCRIPTION
**Pull request by teymour-aldridge**
## 🌟 What is the purpose of this PR?

This migrates the engine from `arrow` (https://github.com/apache/arrow-rs) to `arrow2` (https://github.com/jorgecarleitao/arrow2).

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201893074552404/1201999214936733/f) _(internal)_

## 🔍 What does this change?

- Removes `arrow` and replaces it with `arrow2`. In many cases, the code is now a lot safer than it used to be.

## ⚠️ Known issues

- slower expected simulation run executions

## 🛡 What tests cover this?

- The test suite

### CODE REVIEW

**Comment by nonparibus** [comment]
Would love to see performance benchmarks between the two (although accept we're committed to this now anyway!)

**Comment by nonparibus** [comment]
@teymour-aldridge I've tagged Tim and Alfie in as reviewers now this has been marked ready. If you'd prefer they hold, please Slack.

File: packages/engine/lib/execution/Cargo.toml [hunk 1]
```diff
@@ -22,10 +22,17 @@ rayon = "1.5.3"
...
 serde_json = "1.0.82"
 thiserror = "1.0.31"
-tokio = { version = "1.19.2", features = ["macros", "rt", "sync", "process", "time"] }
+tokio = { version = "1.19.2", features = [
+    "macros",
+    "rt",
+    "sync",
+    "process",
+    "time",
+] }
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/Cargo.toml:31]: I prefer the one-line-version. I guess this is VSCode formatting this?

File: packages/engine/lib/execution/src/package/simulation/context/agent_messages/mod.rs [hunk 2]
```diff
@@ -117,18 +119,34 @@ impl ContextPackage for AgentMessages {
...
+        let mut messages_builder: MutableListArray<
+            i32,
+            MutableFixedSizeListArray<MutablePrimitiveArray<u32>>,
+        > = MutableListArray::new_from(
+            loc_builder,
+            DataType::List(Box::new(Field::new(
+                "item",
+                DataType::FixedSizeList(Box::new(Field::new("item", DataType::UInt32, true)), 3),
+                true,
+            ))),
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/src/package/simulation/context/agent_messages/mod.rs:133]: In `arrow-rs`, `"item"` was hardcoded. Is this still the case with `arrow2`? If no, we probably want to name it properly (can be done in a follow-up).

File: packages/engine/lib/execution/src/package/simulation/state/behavior_execution/package.py [hunk 3]
```diff
@@ -92,24 +92,25 @@ def _format_behavior_error(behavior_name, exc_info):
...
-        #       code.
-
-        if isinstance(m["to"], str):
-            m["to"] = [m["to"]]
-        elif type(m["to"]) != list:
-            raise TypeError(f"Message `to` field must be a list or a string: {m}")
-
-        if not isinstance(m["type"], str):
-            raise TypeError(f"Message `type` field must be a string: {m}")
+    if msgs:
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/src/package/simulation/state/behavior_execution/package.py:95]: `if msgs is not None`?

File: packages/engine/lib/execution/src/package/simulation/state/behavior_execution/reset_index_col.rs [hunk 4]
```diff
@@ -16,16 +17,14 @@ pub struct ResetIndexCol {
 
 impl IntoArrowChange for ResetIndexCol {
     fn get_arrow_change(&self, range: std::ops::Range<usize>) -> memory::Result<ColumnChange> {
+        trace!("resetting index column");
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/src/package/simulation/state/behavior_execution/reset_index_col.rs:20]: How verbose is this line? I don't mind keeping it, just want to flag that currently we misusing `trace` in other places, where `trace!` actually should be `debug!`. I just worry that this is too verbose to run everything with `RUST_LOG=trace`.

File: packages/engine/lib/execution/src/runner/python/state.py [hunk 5]
```diff
@@ -34,8 +34,9 @@ def __getattr__(self, field):  # Can raise AttributeError.
         if field == "messages":
             messages = self.__dict__["__msgs"][idx]
             if not self.__dict__["__msgs_native"][idx]:
-                for message in messages:
-                    message["data"] = json.loads(message["data"])
+                if messages:
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/src/runner/python/state.py:37]: `if messages is not None`?

File: packages/engine/lib/execution/src/runner/python/state.py [hunk 6]
```diff
@@ -170,10 +175,11 @@ def flush_changes(self, schema):
         group_msgs = self.__msg_batch.cols["messages"]
         for i_agent, agent_msgs in enumerate(group_msgs):
             if self.__msgs_native[i_agent]:
-                for msg in agent_msgs:
-                    # When sending a remove-agent message, `data` may be empty to remove self
-                    if "data" in msg:
-                        msg["data"] = json.dumps(msg["data"])
+                if agent_msgs:
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/src/runner/python/state.py:178]: `if agent_messages is not None`?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Partial review - `execution` crate is done. Looks very good and promising so far!

File: packages/engine/lib/memory/src/arrow/batch/mod.rs [hunk 7]
```diff
@@ -99,36 +100,40 @@ impl ArrowBatch {
 
     /// Reload record batch (without checking metaversions).
     pub fn reload_record_batch(&mut self) -> Result<()> {
+        tracing::debug!("reloading record batch");
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/batch/mod.rs:103]: Basically same question as above

File: packages/engine/lib/memory/src/arrow/buffer.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+use num::{Integer, Num};
+
+use super::util::bit_util;
+
+// `n_bits` 0 bits, possibly followed by more 0 bits for padding.
+// todo: we can remove this by using the (less error-prone) `MutableBitmap` instead
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/buffer.rs:6]: Do we have an Asana task for tracking?

File: packages/engine/lib/memory/src/arrow/flush/mod.rs [hunk 9]
```diff
@@ -1,38 +1,48 @@
...
+    #[allow(clippy::redundant_allocation)]
     fn buffer(&self, index: usize) -> &[u8];
     /// Arrow stores the null buffer separately from other buffers.
     fn non_null_buffer_count(&self) -> usize;
+    /// This returns the data of the child arrays.
     fn child_data(&self) -> &[Self];
 }
 
-impl GrowableArrayData for ArrayData {
+impl GrowableArrayData for arrow::array::ArrayData {
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/flush/mod.rs:35]: Do we expect to remove this? I think this was very helpful for testing the new implementation, but is there a benefit in keeping it?

File: packages/engine/lib/memory/src/arrow/flush/mod.rs [hunk 10]
```diff
@@ -179,14 +188,15 @@ pub trait GrowableBatch<D: GrowableArrayData, C: GrowableColumn<D>> {
...
+                // separate
                 debug_assert_eq!(
                     meta.buffer_counts[i] - 1,
                     array_data.non_null_buffer_count(),
                     "Number of buffers in metadata does not match actual number of buffers"
                 );
-                // todo: when adding datatypes with no null buffer (the null datatype), then this
-                //   convention does not work
+                // todo: when adding datatypes with no null buffer (the null datatype), then
+                // this   convention does not work
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/flush/mod.rs:199]:
```suggestion
                // TODO: when adding datatypes with no null buffer (the null datatype), then this
                //   convention does not work
```

File: packages/engine/lib/memory/src/arrow/ipc/offset.rs [hunk 11]
```diff
@@ -0,0 +1 @@
+
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/ipc/offset.rs:1]: Empty file

File: packages/engine/lib/memory/src/arrow/ipc/schema.rs [hunk 12]
```diff
@@ -0,0 +1 @@
+
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/ipc/schema.rs:1]: Empty file

File: packages/engine/lib/memory/src/arrow/ipc/serialize/LICENSE [hunk 13]
```diff
@@ -0,0 +1,2 @@
+This code is licensed under the Apache License (v2.0). A copy of the license may be found at
+https://github.com/jorgecarleitao/arrow2/blob/main/LICENSE
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/ipc/serialize/LICENSE:2]: Could you add a comment to the functions copied in this directory and state what was added/changed? Similar to our old IPC module.

File: packages/engine/lib/memory/src/arrow/mod.rs [hunk 14]
```diff
@@ -2,22 +2,35 @@
...
+//!
+//! # What _is_ Arrow?
+//!
+//! Arrow is an in-memory format for storing columnar data. There are already a lot of good
+//! resources explaining how Arrow works, so we won't duplicate those explanations here, and instead
+//! just list them
+//!
+//! - [Arrow2's guide](https://github.com/DataEngineeringLabs/arrow2-convert)
+//! - [Notes on Arrow](https://wesm.github.io/arrow-site-test/)
+//! - [The Arrow specification](https://arrow.apache.org/docs/format/Columnar.html)
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/mod.rs:14]: Good thing we have the links in place! 👍🏻

File: packages/engine/lib/memory/src/arrow/util/pad.rs [hunk 15]
```diff
@@ -0,0 +1 @@
+pub fn pad_to_8() {}
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/util/pad.rs:1]: What's the reason behind this function?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Another partial review on the `memory` crate. I probably want to take another look at the `ipc` module.

File: packages/engine/lib/simulation-control/Cargo.toml [hunk 16]
```diff
@@ -11,15 +11,22 @@ stateful = { path = "../stateful", default-features = false }
...
 serde_json = "1.0.82"
 thiserror = "1.0.31"
-tokio = { version = "1.19.2", features = ["macros", "rt-multi-thread", "sync", "process", "time"] }
+tokio = { version = "1.19.2", features = [
+    "macros",
+    "rt-multi-thread",
+    "sync",
+    "process",
+    "time",
+] }
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/Cargo.toml:27]: Same as above, let's keep it in one line

File: packages/engine/lib/simulation-control/src/command/create_remove/migration.rs [hunk 17]
```diff
@@ -588,11 +618,11 @@ impl<'a> BufferActions<'a> {
                         let maybe_buffer = if i == 0 {
                             ad.null_buffer()
                         } else {
-                            Some(&ad.buffers()[i - 1])
+                            Some(ad.buffer(i - 1))
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/src/command/create_remove/migration.rs:621]: `ad` = `agent_data`? Can we rename this as a drive-by so it's more readable? Also applies to other actions below.

File: packages/engine/lib/simulation-control/src/command/create_remove/migration.rs [hunk 18]
```diff
@@ -1484,7 +1513,7 @@ pub(super) mod test {
             &agent_schema,
             MemoryId::new(experiment_id.as_uuid()),
         )?;
-        println!("Thru JSON took: {} us", now.elapsed().as_micros());
+        println!("JSON took: {} us", now.elapsed().as_micros());
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/src/command/create_remove/migration.rs:1516]: I guess this was supposed to be removed?

File: packages/engine/lib/simulation-control/src/controller/packages.rs [hunk 19]
```diff
@@ -239,31 +240,44 @@ impl Packages {
...
+                    debug_assert_eq!(arrow_field.data_type(), col.data_type(), "the datatype for {} does not match the schema", arrow_field.name);
+                    if col.len() != num_agents {
+                        panic!(
+                            r#"the length ({}) of the column "{}" does not equal the number of agents \
+                             ({num_agents}). {:#?}"#,
+                            col.len(),
+                            arrow_field.name,
+                            col
+                        )
+                    }
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/src/controller/packages.rs:273]: This can be an `assert_eq!`

File: packages/engine/lib/simulation-control/src/engine.rs [hunk 20]
```diff
@@ -274,6 +274,11 @@ impl Engine {
     fn handle_messages(&mut self, state: &mut State, message_map: &MessageMap) -> Result<()> {
         let message_proxies = state.message_pool().read_proxies()?;
         let mut commands = Commands::from_hash_messages(message_map, &message_proxies)?;
+        // it is necessary to drop `message_proxies` here, because it contains a strong reference to
+        // the MessageBatch; if this strong reference exists, then it is not possible to call
+        // `state.read()` later (because `state.read()` will fail it there are multiple references
+        // to the batch).
+        drop(message_proxies);
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/src/engine.rs:281]: Can this be rewritten or will the borrow checker complain?
```suggestion
        let mut commands = {
            // it is necessary to drop `message_proxies` after reading the commands because it contains a strong reference to
            // the `MessageBatch`; if this strong reference exists, then it is not possible to call
            // `state.read()` later (because `state.read()` will fail it there are multiple references
            // to the batch).
            let message_proxies = state.message_pool().read_proxies()?;
            Commands::from_hash_messages(message_map, &message_proxies)
        };
```

(Will most probably mess up rustfmt)

File: packages/engine/lib/simulation-control/src/tests/arrow.rs [hunk 21]
```diff
@@ -55,30 +55,24 @@ fn get_schema() -> Result<()> {
 
     field_spec_map.try_extend([root_field_spec_from_agent_field(AgentStateField::AgentId)?])?;
 
-    let mut meta = HashMap::new();
+    let mut meta = BTreeMap::new();
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/src/tests/arrow.rs:58]: Any specific reason why choosing `BTreeMap` over `HashMap` here?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Partial review: `simulation-control` crate

File: packages/engine/lib/stateful/src/agent/arrow/array.rs [hunk 22]
```diff
@@ -76,11 +74,11 @@ impl IntoRecordBatch for &[&Agent] {
     }
 
     fn to_agent_batch(&self, schema: &AgentSchema) -> Result<RecordBatch> {
-        let mut cols = Vec::with_capacity(schema.arrow.fields().len());
+        let mut cols = Vec::with_capacity(schema.arrow.fields.len());
 
-        for field in schema.arrow.fields() {
-            // If `name` isn't cloned, Rust wants schema to have longer lifetime.
-            let name = field.name().clone();
+        for field in schema.arrow.fields.clone() {
```
Comment (TimDiekmann) [inline packages/engine/lib/stateful/src/agent/arrow/array.rs:79]: What is the type of `fields`, `Vec`? Is cloning needed here or can it be done by reference?

File: packages/engine/lib/stateful/src/message/pool.rs [hunk 23]
```diff
@@ -114,19 +114,32 @@ impl BatchPool for MessageBatchPool {
 
     fn remove(&mut self, index: usize) -> Self::Batch {
         let batch_arc = self.batches.remove(index);
-        if let Ok(rw_lock) = Arc::try_unwrap(batch_arc) {
+        let res = Arc::try_unwrap(batch_arc);
+        if let Ok(rw_lock) = res {
```
Comment (TimDiekmann) [inline packages/engine/lib/stateful/src/message/pool.rs:118]: I think this can be rewritten as `match`

File: packages/engine/lib/stateful/src/message/pool.rs [hunk 24]
```diff
@@ -114,19 +114,32 @@ impl BatchPool for MessageBatchPool {
...
+                Arc::strong_count(&batch_arc)
+            )
         }
     }
 
     fn swap_remove(&mut self, index: usize) -> Self::Batch {
         let batch_arc = self.batches.swap_remove(index);
-        if let Ok(rw_lock) = Arc::try_unwrap(batch_arc) {
+        let result = Arc::try_unwrap(batch_arc);
+        if let Ok(rw_lock) = result {
```
Comment (TimDiekmann) [inline packages/engine/lib/stateful/src/message/pool.rs:133]: Same as above

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Awesome and impressive work! 🎉 

I finally also completed the `stateful` crate. I noticed a bunch of new `trace!` log calls. I worry a little bit if there are other `trace!` calls that should be `debug!` instead. Generally, I agree with having a more verbose output for tracing.

File: packages/engine/lib/stateful/src/agent/arrow/boolean.rs [hunk 25]
```diff
@@ -43,9 +42,9 @@ impl BooleanColumn {
     }
 }
 
-#[cfg(test)]
+#[cfg(all(test, not(miri)))]
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/stateful/src/agent/arrow/boolean.rs:45]: I think this is fine to ignore for now, but should definitely be addressed sooner than later. Could you add a task for this?

File: packages/engine/lib/stateful/src/agent/arrow/array.rs [hunk 26]
```diff
@@ -76,11 +74,11 @@ impl IntoRecordBatch for &[&Agent] {
...
+            // If `name` isn't cloned, Rust wants schema to have longer lifetime. - DOES IT?
+            let name = field.name.clone();
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/stateful/src/agent/arrow/array.rs]: I think you meant 050254ab34bc785f0449e04fcd3ceb78bd707391 🙂

File: packages/engine/lib/stateful/src/agent/arrow/array.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/stateful/src/agent/arrow/array.rs]: 050254ab34bc785f0449e04fcd3ceb78bd707391

File: packages/engine/lib/simulation-control/src/tests/arrow.rs (hunk 21 above)
Comment (TimDiekmann) [reply packages/engine/lib/simulation-control/src/tests/arrow.rs:58]: That's good enough I guess. Unless we do micro-optimizations it shouldn't matter at all

File: packages/engine/lib/simulation-control/src/engine.rs (hunk 20 above)
Comment (TimDiekmann) [reply packages/engine/lib/simulation-control/src/engine.rs]: Awesome!
Yes, I missed the `?` 😄

File: packages/engine/lib/simulation-control/src/command/create_remove/migration.rs (hunk 18 above)
Comment (TimDiekmann) [reply packages/engine/lib/simulation-control/src/command/create_remove/migration.rs:1518]: Sorry, I mean the `println!` in general. I don't think we want `println!` in our code at all, at least it should be `debug!` or similar.

File: packages/engine/lib/simulation-control/src/command/create_remove/migration.rs (hunk 17 above)
Comment (TimDiekmann) [reply packages/engine/lib/simulation-control/src/command/create_remove/migration.rs]: Thanks, a lot clearer!

File: packages/engine/lib/memory/src/arrow/ipc/serialize/LICENSE (hunk 13 above)
Comment (TimDiekmann) [reply packages/engine/lib/memory/src/arrow/ipc/serialize/LICENSE:2]: Awesome, thank you very much!

File: packages/engine/lib/memory/src/arrow/flush/mod.rs (hunk 9 above)
Comment (TimDiekmann) [reply packages/engine/lib/memory/src/arrow/flush/mod.rs]: Awesome, thanks!

File: packages/engine/lib/execution/Cargo.toml [hunk 27]
```diff
@@ -8,7 +8,7 @@ flatbuffers_gen = { path = "../flatbuffers_gen", default-features = false }
 memory = { path = "../memory", default-features = false }
 stateful = { path = "../stateful", default-features = false }
 
-arrow = { version = "10.0.0", default-features = false }
+arrow2 = {  version = "0.12.0", default-features = false }
```
Comment (TimDiekmann) [inline packages/engine/lib/execution/Cargo.toml:11]:
```suggestion
arrow2 = { version = "0.12.0", default-features = false }
```

File: packages/engine/lib/memory/src/arrow/ipc/mod.rs [hunk 28]
```diff
@@ -0,0 +1,46 @@
+//! This file handles writing Arrow data as raw bytes. We would use the types which [`arrow2`]
```
Comment (TimDiekmann) [inline packages/engine/lib/memory/src/arrow/ipc/mod.rs:1]:
```suggestion
//! This module handles writing Arrow data as raw bytes. We would use the types which [`arrow2`]
```

File: packages/engine/lib/simulation-control/Cargo.toml [hunk 29]
```diff
@@ -11,15 +11,16 @@ stateful = { path = "../stateful", default-features = false }
 execution = { path = "../execution", default-features = false }
 experiment-structure = { path = "../experiment-structure", default-features = false }
 
-arrow = { version = "10.0.0", default-features = false }
+arrow2 = {  version = "0.12.0", default-features = false }
```
Comment (TimDiekmann) [inline packages/engine/lib/simulation-control/Cargo.toml:14]:
```suggestion
arrow2 = { version = "0.12.0", default-features = false }
```

**Review by TimDiekmann — COMMENTED** [review]
Only a few nits 🙂

**Review by TimDiekmann — APPROVED** [review]
AMAZING 🎉 🕺🏻

**Review by TimDiekmann — APPROVED** [review]
